File: LeetCode/BitManipulation/JavaScript/addBinary.js
// Given two binary strings a and b, return their sum as a binary string.

// Example 1:

// Input: a = "11", b = "1"
// Output: "100"
// Example 2:

// Input: a = "1010", b = "1011"
// Output: "10101"

// Constraints:

// 1 <= a.length, b.length <= 104
// a and b consist only of '0' or '1' characters.
// Each string does not contain leading zeros except for the zero itself.

// TC: O(n + m) converting both strings to BigInt and the sum back to base 2
// SC: O(n + m) the BigInt operands and the resulting binary string

const Result = { PASS: "\x1b[92mPASS\x1b[0m", FAIL: "\x1b[91mFAIL\x1b[0m" };

class AddBinaryRecord {
  constructor(a, b, expected) {
    this.a = a;
    this.b = b;
    this.expected = expected;
  }
}

class Solution {
  /**
   * @param {string} a
   * @param {string} b
   * @return {string}
   */
  addBinary(a, b) {
    // Use BigInt instead of parseInt
    let num1 = BigInt(`0b${a}`);
    let num2 = BigInt(`0b${b}`);

    let sum = num1 + num2;

    let binarySum = sum.toString(2);

    return binarySum;

    // Use BigInt instead of parseInt
    // let num1 = BigInt(`0b${a}`);
    // let num2 = BigInt(`0b${b}`);

    // while( num2 !== 0n ) {
    //     let carry = num1 & num2;
    //     num1 = num1 ^ num2;
    //     num2 = carry << 1n;
    // }

    //   return num1.toString(2);
  }
}

function testSolution(record) {
  const solution = new Solution();
  const result = solution.addBinary(record.a, record.b);
  const pass = result === record.expected;

  console.log(`Input: a = ${JSON.stringify(record.a)}, b = ${JSON.stringify(record.b)}`);
  console.log(`Expected: ${JSON.stringify(record.expected)}`);
  console.log(`Result: ${JSON.stringify(result)}`);
  console.log(pass ? Result.PASS : Result.FAIL);
}

const records = [
  new AddBinaryRecord("11", "1", "100"),
  new AddBinaryRecord("1010", "1011", "10101"),
  new AddBinaryRecord("0", "0", "0"),
  new AddBinaryRecord("1", "111", "1000"),
  new AddBinaryRecord("1111", "1111", "11110"),
];

records.forEach((record, index) => {
  console.log(`# Test case ${index + 1}`);
  testSolution(record);
  console.log("----------------------------------------");
});

File: LeetCode/BitManipulation/python/addBinary.py
from dataclasses import dataclass
from enum import Enum
from typing import List


class Result(Enum):
    PASS = "\033[92mPASS\033[00m"
    FAIL = "\033[91mFAIL\033[00m"

@dataclass
class addBinaryRecord:
    a: str
    b: str
    expected: str
    
class Solution:
    def addBinary1(self, a: str, b: str) -> str:
        a: int = int("0b" + a, 2)
        b: int = int("0b" + b, 2)
        return bin(a + b)[2:]
    
    def addBinary2(self, a: str, b: str) -> str:
        i: int = len(a) - 1
        j: int = len(b) - 1
        carry: int = 0

        res: str = ""

        while i > -1 or j > -1 or carry == 1:
            sum: int = carry

            if i > -1:
                sum += int(a[i]) - int('0')
            
            if j > -1:
                sum += int(b[j]) - int('0')

            i -= 1
            j -= 1

            res += str(sum % 2)
            carry = sum // 2
        
        # return "".join(reversed(res))
        return res[::-1]

    @staticmethod
    def testSolution(record: addBinaryRecord) -> None:
        print(f"input:\ta: {record.a}")
        print(f"\tb: {record.b}")
        print(f"expected: {record.expected}")

        res: str = Solution().addBinary1(record.a, record.b)
        print(f"result: {res}")
        print(Result.PASS.value if res == record.expected else Result.FAIL.value)

if __name__ == "__main__":
    records: List[addBinaryRecord] = [
        addBinaryRecord("11", "1", "100"),
        addBinaryRecord("1010", "1011", "10101"),
        addBinaryRecord("10100000100100110110010000010101111011011001101110111111111101000000101111001110001111100001101",
                        "110101001011101110001111100110001010100001101011101010000011011011001011101111001100000011011110011",
                        "110111101100010011000101110110100000011101000101011001000011011000001100011110011010010011000000000")
    ]

    for i, record in enumerate(records):
        print(f"# Test case {i + 1}")
        Solution.testSolution(record)
        print("-" * 50)